### PR TITLE
Replace 16-bit code in MSW wxGetUserName

### PR DESCRIPTION
**Pull request by Blake-Madden**
Uses `NetAPIXXX` functions now instead of `GetProfileString`.
Also reformats user names like "Smart, Julian" into "Julian Smart".

### CODE REVIEW

**Review by vadz — COMMENTED** [review]
Thanks!

This looks globally good, but I think the error handling should be slightly improved and there are a couple of possible simplifications (I didn't mention using `wxDL_INIT_FUNC` as one of them as (not) using macros is at least partially a question of taste, but it could be used to avoid having to repeat the function names multiple times too).

File: src/msw/utils.cpp [hunk 1]
```diff
@@ -258,93 +274,116 @@ bool wxGetUserId(wxChar *buf,
...
-    MultiByteToWideChar( CP_ACP, 0, szUserName, strlen(szUserName)+1,
-            wszUserName, WXSIZEOF(wszUserName) );
-    MultiByteToWideChar( CP_ACP, 0, szDomain, strlen(szDomain)+1,
-            wszDomain, WXSIZEOF(wszDomain) );
+    LPBYTE ComputerName{ nullptr };
+    USER_INFO_2* ui2{ nullptr };
 
-    // Get the computer name of a DC for the domain.
-    if ( NetGetDCName( nullptr, wszDomain, &ComputerName ) != NERR_Success )
+    wxDynamicLibrary netapi32("netapi32");
```
[the remark this answers is not in the document]
Comment (Blake-Madden) [reply src/msw/utils.cpp]: Done

File: src/msw/utils.cpp [hunk 2]
```diff
@@ -258,93 +274,116 @@ bool wxGetUserId(wxChar *buf,
...
+    // filled with the login name.
+    case NERR_InvalidComputer:
+        wxLogWarning(_("Invalid domain controller name."));
+        wxFALLTHROUGH;
+
+    case NERR_UserNotFound:
+        wxLogWarning(_("Invalid user name '%s'."), buf);
+        wxFALLTHROUGH;
+
+    case ERROR_ACCESS_DENIED:
```
[the remark this answers is not in the document]
Comment (Blake-Madden) [reply src/msw/utils.cpp]: Done. Removed `switch` block, uses a simple check for non-success.

File: src/msw/utils.cpp [hunk 3]
```diff
@@ -258,93 +274,116 @@ bool wxGetUserId(wxChar *buf,
...
-
-        default:
-            wxLogSysError(wxT("Can't get information about user"));
-
-            goto error;
+    if (netGetAnyDCName == nullptr ||
+        netUserGetInfo == nullptr ||
+        netApiBufferFree == nullptr)
+    {
+        return true;
```
[the remark this answers is not in the document]
Comment (Blake-Madden) [reply src/msw/utils.cpp]: I changed the logic to return false now.

File: src/msw/utils.cpp [hunk 4]
```diff
@@ -258,93 +274,116 @@ bool wxGetUserId(wxChar *buf,
...
+        wxLogWarning(_("The network path specified in the "
+                       "servername parameter was not found."));
+        wxFALLTHROUGH;
+
+    case ERROR_INVALID_LEVEL:
+        wxLogWarning(_("The value specified for the "
+                       "level parameter is invalid."));
+        wxFALLTHROUGH;
+
+    default:
```
[the remark this answers is not in the document]
Comment (Blake-Madden) [reply src/msw/utils.cpp]: Removed the switch block, just checks for success now.

File: src/msw/utils.cpp [hunk 5]
```diff
@@ -258,93 +274,116 @@ bool wxGetUserId(wxChar *buf,
...
-            break;
-
-        case NERR_InvalidComputer:
-            wxLogError(wxT("Invalid domain controller name."));
+    const NetGetAnyDCName_t netGetAnyDCName =
+        reinterpret_cast<NetGetAnyDCName_t>(netapi32.GetSymbol("NetGetAnyDCName"));
+    const NetUserGetInfo_t netUserGetInfo =
+        reinterpret_cast<NetUserGetInfo_t>(netapi32.GetSymbol("NetUserGetInfo"));
+    const NetApiBufferFree_t netApiBufferFree =
+        reinterpret_cast<NetApiBufferFree_t>(netapi32.GetSymbol("NetApiBufferFree"));
```
[the remark this answers is not in the document]
Comment (Blake-Madden) [reply src/msw/utils.cpp]: Done

File: src/msw/utils.cpp [hunk 6]
```diff
@@ -122,21 +115,44 @@
...
+using NetUserGetInfo_t = NET_API_STATUS(NET_API_FUNCTION*)(
+    LPCWSTR servername,
+    LPCWSTR username,
+    DWORD   level,
+    LPBYTE* bufptr);
+using NetApiBufferFree_t = NET_API_STATUS(NET_API_FUNCTION*)(
+    LPVOID Buffer);
+
+// Calls NetApiBufferFree on a buffer created from NetApiBufferAllocate()
+struct NetApiBufferAutoFree
```
[the remark this answers is not in the document]
Comment (Blake-Madden) [reply src/msw/utils.cpp]: Done. Removed auto deleter class also.

File: src/msw/utils.cpp [hunk 7]
```diff
@@ -258,93 +274,116 @@ bool wxGetUserId(wxChar *buf,
...
+        // In the case of full name being in the format of "[LAST_NAME], [FIRST_NAME]",
+        // reformat it to a more readable "[FIRST_NAME] [LAST_NAME]".
+        const size_t commaPosition = fullUserName.find(", ");
+        if (commaPosition != wxString::npos)
+            {
+            const wxString firstName = fullUserName.substr(commaPosition + 2);
+            fullUserName.erase(commaPosition);
+            fullUserName.insert(0, firstName + ' ');
+            }
+        wxStrncpy(buf, fullUserName.t_str(), maxSize);
```
[the remark this answers is not in the document]
Comment (Blake-Madden) [reply src/msw/utils.cpp]: Done

File: interface/wx/utils.h [hunk 8]
```diff
@@ -856,9 +856,8 @@ wxString wxGetUserHome(const wxString& user = wxEmptyString);
 /**
     This function returns the "user id" also known as "login name" under Unix
     (i.e. something like "jsmith"). It uniquely identifies the current user (on
-    this system).  Under Windows or NT, this function first looks in the
-    environment variables USER and LOGNAME; if neither of these is found, the
-    entry @b UserId in the @b wxWidgets section of the WIN.INI file is tried.
+    this system).  Under Windows, this function looks in the
+    environment variables USER and LOGNAME.
```
Comment (PBfordev) [inline interface/wx/utils.h:860]: Looking at the code, it seems this is not true:
https://github.com/wxWidgets/wxWidgets/blob/ac3cea5bcd5a3fc3d52648efb4dceacbcc9d0698/src/msw/utils.cpp#L243-L259

File: interface/wx/utils.h [hunk 9]
```diff
@@ -883,9 +882,10 @@ bool wxGetUserId(char* buf, int sz);
 /**
     This function returns the full user name (something like "Mr. John Smith").
 
-    Under Windows or NT, this function looks for the entry UserName in the
-    wxWidgets section of the WIN.INI file. If PenWindows is running, the entry
-    Current in the section User of the PENWIN.INI file is used.
+    Under Windows, this function looks for the entry UserName in the
```
Comment (PBfordev) [inline interface/wx/utils.h:885]: The WIN.INI part should be removed (as the useless `::GetProfileString()` was removed) and the rest changed.

BTW, does the function really return "something like "Mr. John Smith" (see the first paragraph), i.e., including stuff such as "Mr. "?

File: interface/wx/utils.h (hunk 8 above)
Comment (Blake-Madden) [reply interface/wx/utils.h:860]: Fixed to say just "USERNAME" now.

File: interface/wx/utils.h (hunk 9 above)
Comment (Blake-Madden) [reply interface/wx/utils.h]: Fixed, and expanded help. No, "Mr. " will not be included (unless you actually put that into the system as part of your first name); I removed that. Thanks!

**Comment by vadz** [comment]
Looks good, thanks! Will (squash) merge soon.

**Pull request closed without merging**